src/tracer.rs

```diff
@@ -148,19 +172,37 @@ pub(crate) unsafe fn parent(
             (true, WaitStatus::PtraceSyscall(_)) => {
                 let cur_info = children.get(&pid).unwrap(); // it's guaranteed here that get() returns Some
                 let started_syscall = !cur_info.in_syscall;
+                let is_spoiled = cur_info.in_spoiled;
+                let spoiled_syscall = cur_info.spoiled_syscall;
                 let new_info = ChildInfo {
                     in_syscall: started_syscall,
+                    in_spoiled: false,
+                    spoiled_syscall: 0,
                 };
                 children.insert(pid, new_info);
-                let regs = nix::sys::ptrace::getregs(Pid::from_raw(pid as i32))
+                let mut regs = nix::sys::ptrace::getregs(Pid::from_raw(pid as i32))
                     .context("ptrace getregs failed")?;
-                let params = decode_syscall_args(regs);
+                if is_spoiled {
+                    return_eio(Pid::from_raw(pid as i32), &mut regs, spoiled_syscall).context("return -EIO failed")?;
+                }
+                let params = decode_syscall_args(&regs);
                 let def = magic.lookup_syscall_by_id(SyscallId(params.syscall_id as u32));
                 let child_pid = Pid::from_raw(pid as i32);
                 let mut decoded_params = match def {
                     Some(def) => process_syscall(&params, child_pid, magic, def),
                     None => None,
                 };
+                if started_syscall && decoded_params.is_some() && (params.syscall_id == 2 /*open*/ || params.syscall_id == 257 /*openat*/) { //TODO: x86_64 only
```

Comment:
Add some comment that describes what this if does.
Also, you should skip this entire `if` if fail_path is disabled:

```rust
if let Some(fail_path) = &setting.fail_path {
   // ...
}

```